compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,355 @@ fn check_for_underconstrained_values_within_function(
+        // one of the results should be constrained
+        for (i, result_status) in self.results.iter().enumerate() {
+            match result_status {
+                // Skip checking already covered results
+                ResultStatus::Constrained => {
+                    continue;
+                }
+                ResultStatus::Unconstrained { descendants } => {
+                    if descendants.intersection(constrained_values).next().is_some() {
+                        results_involved.push(i);
+                    }
+                }
+            }
+        }
+
+        // Also, one of the argument descendants should be constrained
+        // (skipped if there were no arguments, or if an actual result and not a
+        // descendant has been constrained, e.g. against a constant)
+        if !results_involved.is_empty()
+            && (self.arguments.is_empty()
+                || self.root_results.intersection(constrained_values).next().is_some()
+                || self.arguments.intersection(constrained_values).next().is_some())
+        {
+            // Remember the partial constraint, clearing the sets
+            results_involved.iter().for_each(|i| self.results[*i] = ResultStatus::Constrained);
+        }
+    }
+}
+
+impl DependencyContext {
+    /// Build the dependency context of variable ValueIds, storing
+    /// information on value ids involved in unchecked Brillig calls
+    fn build(&mut self, function: &Function, all_functions: &BTreeMap<FunctionId, Function>) {
+        self.block_queue.push(function.entry_block());
+        while let Some(block) = self.block_queue.pop() {
+            if self.visited_blocks.contains(&block) {
+                continue;
+            }
+            self.visited_blocks.insert(block);
+            self.process_instructions(block, function, all_functions);
+        }
+    }
+
+    /// Go over the given block tracking Brillig calls and checking them against
+    /// following constraints
+    fn process_instructions(
+        &mut self,
+        block: BasicBlockId,
+        function: &Function,
+        all_functions: &BTreeMap<FunctionId, Function>,
+    ) {
+        trace!("processing instructions of block {} of function {}", block, function);
+
+        for instruction in function.dfg[block].instructions() {
+            // Optimization: there is no reason to do anything until a Brillig
+            // call is encountered
+            if self.tainted.is_empty() {
+                let mut brillig_encountered = false;
+                if let Instruction::Call { func: func_id, .. } = &function.dfg[*instruction] {
+                    if let Value::Function(callee) = &function.dfg[*func_id] {
+                        if all_functions[&callee].runtime().is_brillig() {
+                            brillig_encountered = true;
+                        }
+                    }
+                }
+                if !brillig_encountered {
+                    continue;
```

Comment:
What would happen in this example?
```rust
fn main(a: Field, b: Field) {
  let c = a + b;
  let d = unsafe { foo(a, b) };
  assert_eq(c, d);
}

unconstrained fn add(a: Field, b: Field) -> Field { 
  a + b
}
```

Reply:
Hmm. Good point. Perhaps this optimization is not a good idea, at least being as sweeping. Removing it for now

Reply:
I'm not sure that solves the issue. 

To be clear what we see here is that `c` is a descendant of `a` and `b`, so a constraint on `c` puts a constraint on `a` and `b`, which are the arguments of `add` and `sub`, and the `assert_eq` puts `c` in relation with the outputs of both brillig calls, and they aren't direct descendants. 

The problem is that even if you remove the optimisation, the only mechanism to remember that `c` had anything to do with `a` and `b` is the `tainted` collection, ie. until we see a call to `add` and `sub` we forget how `c` came to be, even if we process the instruction.

Reply:
Maybe we should do two passes:
1. one to collect all descendants of any Brillig call, then 
2. to collect all ancestors of constraints that intersect with that.

Reply:
That's true. Probably could go backwards from the calls found unconstrained by the end only. I think this should go into a possible false positive issue like EnableSideEffectsIf though. Also, could lead to a significant performance hit, especially for huge unrolled functions.

Reply:
Moved into a separate issue: https://github.com/noir-lang/noir/issues/6735